### PR TITLE
Snapshotter

### PR DESCRIPTION
**Pull request by evil159**
### What does this pull request do?

This PR exposes `Snapshotter` and introduces refactored event handing for `MapWidget` that is also reused for `Snapshotter`. Event listeners(map loaded, style loaded, map idle etc.) are no longer being ignored after `MapWidget` state initialization, they can be updated dynamically.

Based https://github.com/mapbox/mapbox-maps-flutter/pull/386.
Addresses https://github.com/mapbox/mapbox-maps-flutter/issues/210, https://github.com/mapbox/mapbox-maps-flutter/issues/279, https://mapbox.atlassian.net/browse/MAPSFLT-159

### What is the motivation and context behind this change?

`Snapshotter` provides a way to get an image snapshot of a map without instantiating a `MapWidget`

### Pull request checklist:
 - [x] Add a changelog entry.
 - [x] Write tests for all new functionality. If tests were not written, please explain why.
 - [x] Add documentation comments for any added or updated public APIs.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -15,6 +15,31 @@ mapboxMap?.location.updateSettings(LocationComponentSettings(
...
+
+```dart
+final snapshotter = await Snapshotter.create(
+  options: MapSnapshotOptions(
+      size: Size(width: 400, height: 400),
+      pixelRatio: MediaQuery.of(context).devicePixelRatio),
+  onStyleLoadedListener: (_) {
+    // apply runtime styleing
+    final layer = CircleLayer(id: "circle-layer", sourceId: "poi-source");
+    _snapshotter?.style.addLayer(layer);
```
Comment (pjleonard37) [inline CHANGELOG.md:34]: Does this need the _?

```suggestion
    snapshotter?.style.addLayer(layer);
```

File: example/lib/snapshotter.dart [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+  _onMapCreated(MapboxMap mapboxMap) async {
+    this.mapboxMap = mapboxMap;
+
+    _snapshotter = await Snapshotter.create(
+      options: MapSnapshotOptions(
+          size: Size(width: 400, height: 400),
+          pixelRatio: MediaQuery.of(context).devicePixelRatio),
+    );
+    await _snapshotter?.style.setStyleURI(MapboxStyles.STANDARD);
```
Comment (pjleonard37) [inline example/lib/snapshotter.dart:47]: Perhaps a different style to show the difference: 

```suggestion
    await _snapshotter?.style.setStyleURI(MapboxStyles.OUTDOORS);
```

File: ios/Classes/Extensions.swift [hunk 3]
```diff
@@ -716,3 +786,28 @@ extension Optional {
         lhs = rhs
     }
 }
+
```
Comment (pjleonard37) [inline ios/Classes/Extensions.swift:789]: Nit: delete completely

**Review by pjleonard37 — APPROVED** [review]

**Comment by pjleonard37** [comment]
Looks great, @evil159! 📸

File: lib/src/map_widget.dart [hunk 4]
```diff
@@ -243,39 +208,43 @@ class _MapWidgetState extends State<MapWidget> {
...
-      onRenderFrameFinishedListener: widget.onRenderFrameFinishedListener,
-      onRenderFrameStartedListener: widget.onRenderFrameStartedListener,
-      onSourceAddedListener: widget.onSourceAddedListener,
-      onSourceDataLoadedListener: widget.onSourceDataLoadedListener,
-      onSourceRemovedListener: widget.onSourceRemovedListener,
-      onStyleDataLoadedListener: widget.onStyleDataLoadedListener,
-      onStyleImageMissingListener: widget.onStyleImageMissingListener,
-      onStyleImageUnusedListener: widget.onStyleImageUnusedListener,
-      onResourceRequestListener: widget.onResourceRequestListener,
-      onMapTapListener: widget.onTapListener,
```
Comment (bf-vs) [inline lib/src/map_widget.dart:271]: `onMapTapListener` does no longer works in our app after this commit. Should we migrate to something else. If so what is it?

Comment (evil159) [reply lib/src/map_widget.dart:271]: Thanks for the heads up, this was removed accidentally, it is restored in https://github.com/mapbox/mapbox-maps-flutter/pull/526

**Pull request merged**